File: 프로그래머스/level2/순위검색.cpp
#include <string>
#include <map>
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

map<string, vector<int>> mp;

vector<string> divide(string s){
    string word = "";
    vector<string> entity;
    for(int i=0; i<s.length(); i++){
        if(s[i]==' '){
            entity.push_back(word);
            word = "";
        }
        else{
            word += s[i];
        }
    }
    entity.push_back(word);
    return entity;
}
string find_now(string tmp){
    string now;
    if(tmp[0]=='c')       now = "1";
    else if(tmp[0]=='j')  now = "2";
    else if(tmp.compare("python")==0)  now = "3";
    else if(tmp[0]=='b')  now = "1";
    else if(tmp[0]=='f')  now = "2";
    else if(tmp[0]=='s')  now = "1";
    else if(tmp[0]=='j')  now = "2";
    else if(tmp[0]=='p')  now = "2";
    else if(tmp[0]=='a')  now = "";
    else                        now = "0";
    return now;
}
void classfy(string info){
    vector<string> entity;
    entity = divide(info);
    int value = stoi(entity[entity.size()-1]);
    entity.pop_back();
    vector<string> input;
    input.push_back("");
    for(int i=0; i<entity.size(); i++){
        string now = "";
        vector<string> tmp;
        now = find_now(entity[i]);
        for(int j=0; j<input.size(); j++){
            tmp.push_back(input[j]+now);
            tmp.push_back(input[j]+"0");
        }
        input = tmp;
    }
    for(int i=0; i<input.size(); i++){
        mp[input[i]].push_back(value);
    }
}
void find(string query, vector<int> &answer){
    vector<string> entity;
    entity = divide(query);
    int value = stoi(entity[entity.size()-1]);
    entity.pop_back();
    string now = "";
    for(int i=0; i<entity.size(); i++){
        now += find_now(entity[i]);
    }
    int s = 0;
    int e = mp[now].size()-1;
    int target=mp[now].size();
    while(s<=e){
        int mid = (s+e)/2;
        if(mp[now][mid]>=value){
            e = mid-1;
            target = mid;
        }
        else    s = mid+1;
    }
    answer.push_back(mp[now].size()-target);
}
vector<int> solution(vector<string> info, vector<string> query) {
    vector<int> answer;
    for(int i=0; i<info.size(); i++){
        classfy(info[i]);
    }
    map<string, vector<int>> ::iterator it;
    for(it=mp.begin(); it!=mp.end(); it++){
        sort((*it).second.begin(), (*it).second.end());
    }
    for(int i=0; i<query.size(); i++){
        find(query[i], answer);
    }
    return answer;
}

File: 프로그래머스/level2/순위검색.js
const classfy = (str) => {
    let arr = str.split(" ");
    let ans = [];
    for(let i=0; i<4; i++){
        if(arr[i]==="python")   ans.push("3");
        else if(arr[i]==="java")   ans.push("2");
        else if(arr[i][0] === 'c'||arr[i][0] === 'b'||arr[i][0] === 'j')  ans.push("1");
        else if(arr[i][0] === 'f'||arr[i][0] === 's'||arr[i][0] === 'p')  ans.push("2");
    }
    ans.push(Number(arr[4]));
    return ans;
}

function solution(info, query) {
    var answer = [];
    let obj = {};
    info.forEach(str => {
        let tmp = [];
        let arr = classfy(str);
        for(let i=0; i<4; i++){
            let cp = tmp.slice();
            tmp = [];
            if(cp.length===0){
                tmp.push(arr[i]);
                tmp.push("0");
            }
            else{
                for(let j=0; j<cp.length; j++){
                    tmp.push(cp[j] + arr[i]);
                    tmp.push(cp[j] + "0");
                }
            }
        }
        tmp.forEach(str => {
            if(obj[str] === undefined){
                obj[str] = [arr[4]];
            }
            else{
                obj[str].push(arr[4]);   
            }
        });
    });
    for(const now in obj){
        obj[now] = obj[now].sort((a,b)=>a-b);
    }
    query.forEach(q => {
        let tmp = q.split(" ");
        let arr = [];
        for(let i=0; i<=6; i+=2){
            arr.push(tmp[i]);
        }
        let num = Number(tmp[7]);
        let str = "";
        for(let i=0; i<arr.length; i++){
            if(arr[i]==="python")   str += "3";
            else if(arr[i]==="java")   str += "2";
            else if(arr[i][0] === 'c'||arr[i][0] === 'b'||arr[i][0] === 'j')  str += "1";
            else if(arr[i][0] === 'f'||arr[i][0] === 's'||arr[i][0] === 'p')  str += "2";
            else if(arr[i]==='-')   str+="0";
        }
        if(obj[str] === undefined){
            answer.push(0);
        }
        else{
            let idx = obj[str].length;
            let s = 0;
            let e = obj[str].length;
            while(s<=e){
                let mid = Math.floor((s+e)/2);
                if(obj[str][mid]>=num){
                    idx = mid;
                    e = mid - 1;
                }
                else{
                    s = mid + 1;
                }
            }
            answer.push(obj[str].length-idx);
        }
        
    });
    return answer;
}